getting started diary example not working

Hello. Trying to get started with crypton. Following the getting started guide. I'm stuck on "running the diary example"

I'm guessing I need to run `crypton-server db:init`; but get error: `no such file or directory, open './server/config/config.development.json'`

Also tried `crypton-server run`, but get error: `password authentication failed for user "crypton_development_user"`

Any assistance is appreciated.

I was on node v5.5, which isn't supported. Switched back to 4.2.6. Ran make from top level. Server tests pass. One client test fails: 

```
PhantomJS 1.9.8 (Linux 0.0.0) Core default properties should have the correct version FAILED
	AssertionError: expected '0.1.0' to equal '0.0.4'
```

But it is also failing on travis, so not sure that is the problem.

When run integration tests, after all the setup, I get:

```
...
Compiling client code to crypton.js...
Starting Crypton backend...
Running integration tests...
Killing Crypton backend...
```

which suggests that maybe the integration tests aren't running?

after all this, still get the same `no such file or directory, open './server/config/config.development.json'` when running `crypton-server db:init`. But also not sure that's the command I should be running.

Thanks for your help here. I'll make a PR to clarify the getting started guide once this is all working.

--- a/README.md
+++ b/README.md
@@ -26,20 +26,24 @@ wget http://download.redis.io/redis-stable.tar.gz
 tar xvzf redis-stable.tar.gz
 cd redis-stable
 make
-
-sudo cp src/redis-server /usr/local/bin/
-sudo cp src/redis-cli /usr/local/bin/
+make test
+sudo make install
 ```
+
+Configuring and setting up redis to run is an environment specific problem
+and left as an exercise for the user. You may now be able to just simply
+do `redis-server &`
 
 ### Install PostgreSQL Database
 
-[PostgreSQL](http://www.postgresql.org/download/linux/) supports a variety of Linux versions:
+[PostgreSQL](http://www.postgresql.org/download/linux/) supports a variety of
+Linux versions. One of these install commands may work for you:
 
 - [Red Hat](http://www.postgresql.org/download/linux/redhat/): `yum install postgresql-server`
 
 - [Debian](http://www.postgresql.org/download/linux/debian/): `apt-get install postgresql-9.4`
 
-- [Ubuntu](http://www.postgresql.org/download/linux/ubuntu/): `apt-get install postgresql-9.4`
+- [Ubuntu](http://www.postgresql.org/download/linux/ubuntu/): `sudo apt-get install postgresql postgresql-contrib`
 
 - [SuSE](http://www.postgresql.org/download/linux/suse/): See the [official instructions](http://www.postgresql.org/download/linux/suse/) for more detail
 
@@ -66,22 +70,27 @@ After installation follow the instructions to copy the `launchd` plist file and 
 
 #### Create a new `postgres` database user using the `createuser` command.
 
-`createuser` is installed as part of the `postgresql` install.
+`createuser` is installed as part of the `postgresql` install. This command will
+create a new user called `postgres` that has superuser role permissions and
+can create roles.
 
 ````
-createuser -s -r postgres
+createuser --superuser --createrole postgres
 ````
 
 ## OS Common
 
 ### Install the [Node.js Version Manager](https://github.com/creationix/nvm) and Node.js
 
-See the [nvm](https://github.com/creationix/nvm) website for detailed installation instructions. Crypton currently expects Node.js version `4.2.1` to be installed which is the most current LTS release.
+See the [nvm](https://github.com/creationix/nvm) website for detailed installation instructions. Crypton currently expects Node.js version `4.2.2` to be installed which is the most current LTS release.
 
 ````
-curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.29.0/install.sh | bash
+curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.30.2/install.sh | bash
 
 nvm install 4.2.2
+
+# optional - set this as your default node install
+nvm alias default v4.2.2
 ````
 
 Verify your `node` installation:
@@ -114,6 +123,46 @@ You need to do this ***every*** time you `cd` into the `crypton` dir in every te
 npm install -g karma-cli
 ````
 
+## Configure your master postgres database user:
+
+You need to configure your environment specific config files for crypton to use
+the postgres database user that has superuser permissions. Edit the following
+files:
+
+```
+server/config/config.development.json
+server/config/config.test.json
+
+# create as needed from a copy of the other config files:
+server/config/config.production.json
+```
+
+In those files you need to change the `postgres_user` and `postgres_user_password`
+keys as appropriate for your environment. The user you configure must have
+postgres superuser permissions.
+
+e.g.
+```
+"database": {
+   "type": "postgres",
+   "postgres_user": "postgres",        <-- CHANGE AS NEEDED
+   "postgres_user_password": "",       <-- CHANGE AS NEEDED
+   "user": "crypton_development_user",
+   "password": "crypton_development_user_password",
+   "database": "crypton_development",
+   "host": "localhost",
+   "port": 5432
+ },
+ ...
+```
+
+The default is username `postgres` with an empty password.
+
+NOTE: If you are on Ubuntu Linux please note that you must have a username
+and password set for a postgres superuser and you need to ensure you can connect to the DB with `psql -U username`. You need to be able to do this
+without using `sudo` to first shell out to the `postgres` user that is created by
+the default installation scripts on Ubuntu. [This page might be of interest](https://stackoverflow.com/questions/7695962/postgresql-password-authentication-failed-for-user-postgres).
+
 ## Test that all Crypton dependencies are installed
 
 ````
@@ -135,7 +184,7 @@ npm install
 
 ## Build the Crypton Node.js Server and setup the DB
 
-The `db:init` command creates a `crypton_test_user` DB user, `crypton_test` database, and the appropriate DB schema in PostgreSQL.
+The `db:init` command creates a `crypton_ENVIRONMENT_user` DB user, `crypton_ENVIRONMENT` database, and the appropriate DB schema in PostgreSQL based on your `NODE_ENV`.
 
 ````
 cd crypton/server

--- a/check_dependencies.sh
+++ b/check_dependencies.sh
@@ -44,8 +44,8 @@ if [ $POSTGRES_USER_EXISTS ]; then
   echo "Found postgres user..."
 else
   echo "Configuration issue: the postgres user does not exist in the database, please create it:"
-  echo "  createuser -s -r postgres"
-  echo "is one way to create the postgres user"
+  echo "  createuser --superuser --createrole postgres"
+  echo "is one possible way to create the postgres superuser"
   exit 1;
 fi
 

--- a/server/bin/drop.js
+++ b/server/bin/drop.js
@@ -1,13 +1,24 @@
 var pg = require('pg').native;
-var config = require('../lib/config');
 var fs = require('fs');
-var realConfig = JSON.parse(fs.readFileSync('./server/config/config.' + process.env.NODE_ENV + '.json', 'utf8'));
-var dbConfig = realConfig.database;
+
+var envConfig = JSON.parse(fs.readFileSync(__dirname + '/../config/config.' + process.env.NODE_ENV + '.json', 'utf8'));
+
+// Clone the config and use the master postgres account for this server to
+// create the new user and DB we need, and then revert back to
+// using the env specific user and password for pg.
+// These values come from 'GIT_ROOT/server/config/config.ENVIRONMENT.json'
+var postgresUserConfig = JSON.parse(JSON.stringify(envConfig));
+postgresUserConfig.database.database = 'postgres';
+postgresUserConfig.database.user = postgresUserConfig.database.postgres_user;
+postgresUserConfig.database.password = postgresUserConfig.database.postgres_user_password;
 
 function connect(callback) {
   'use strict';
 
-  pg.connect(config.database, function(err, client, done) {
+  console.log('Using DB config : ', postgresUserConfig.database);
+
+  // For DROP functions, always connecting as the master postgres user.
+  pg.connect(postgresUserConfig.database, function(err, client, done) {
     if (err) {
       console.log(err);
       process.exit(1);
@@ -20,11 +31,11 @@ function connect(callback) {
 function dropDatabase(callback) {
   'use strict';
 
-  console.log('Dropping database...');
+  console.log('Dropping database : ', envConfig.database.database);
 
   connect(function(client, done) {
     var query = {
-      text: 'drop database ' + dbConfig.database,
+      text: 'DROP DATABASE ' + envConfig.database.database,
     };
 
     client.query(query, function(err, result) {
@@ -35,7 +46,7 @@ function dropDatabase(callback) {
         return callback();
       }
 
-      console.log('Dropped database');
+      console.log('Dropped database : ', envConfig.database.database);
       callback();
     });
   });
@@ -44,11 +55,11 @@ function dropDatabase(callback) {
 function dropUser(callback) {
   'use strict';
 
-  console.log('Dropping user...');
+  console.log('Dropping user : ', envConfig.database.user);
 
   connect(function(client, done) {
     var query = {
-      text: 'drop role ' + dbConfig.user,
+      text: 'DROP ROLE ' + envConfig.database.user,
     };
 
     client.query(query, function(err, result) {
@@ -59,7 +70,7 @@ function dropUser(callback) {
         return callback();
       }
 
-      console.log('Dropped user');
+      console.log('Dropped user : ', envConfig.database.user);
       callback();
     });
   });
@@ -68,12 +79,9 @@ function dropUser(callback) {
 module.exports = function() {
   'use strict';
 
-  config.database.user = 'postgres';
-  config.database.database = 'postgres';
-
   dropDatabase(function() {
     dropUser(function() {
-      console.log('Done');
+      console.log('Done dropping database and user.');
       process.exit();
     });
   });

--- a/server/bin/init.js
+++ b/server/bin/init.js
@@ -1,14 +1,27 @@
 var fs = require('fs');
 var pg = require('pg').native;
-var config = require('../lib/config');
-var oldConfig;
-var realConfig = JSON.parse(fs.readFileSync('./server/config/config.' + process.env.NODE_ENV + '.json', 'utf8'));
-var dbConfig = realConfig.database;
+
+// will equal envConfig or postgresUserConfig depending on context
+var currentConfig;
+
+var envConfig = JSON.parse(fs.readFileSync(__dirname + '/../config/config.' + process.env.NODE_ENV + '.json', 'utf8'));
+
+// Clone the config and use the master postgres account for this server to
+// create the new user and DB we need, and then revert back to
+// using the env specific user and password for pg.
+// These values come from 'GIT_ROOT/server/config/config.ENVIRONMENT.json'
+var postgresUserConfig = JSON.parse(JSON.stringify(envConfig));
+postgresUserConfig.database.database = 'postgres';
+postgresUserConfig.database.user = postgresUserConfig.database.postgres_user;
+postgresUserConfig.database.password = postgresUserConfig.database.postgres_user_password;
 
 function connect(callback) {
   'use strict';
 
-  pg.connect(config.database, function(err, client, done) {
+  console.log("\nUsing DB config: ", currentConfig.database);
+
+  // Will connect with different users depending on what we are doing.
+  pg.connect(currentConfig.database, function(err, client, done) {
     if (err) {
       console.log(err);
       process.exit(1);
@@ -21,11 +34,11 @@ function connect(callback) {
 function createUser(callback) {
   'use strict';
 
-  console.log('Creating user...');
+  console.log('Creating user : ', envConfig.database.user);
 
   connect(function(client, done) {
     var query = {
-      text: 'create user ' + dbConfig.user + ' with encrypted password \'' + dbConfig.password + '\'',
+      text: 'CREATE USER ' + envConfig.database.user + ' WITH ENCRYPTED PASSWORD \'' + envConfig.database.password + '\'',
     };
 
     client.query(query, function(err, result) {
@@ -36,7 +49,7 @@ function createUser(callback) {
         return callback();
       }
 
-      console.log('User created');
+      console.log('Created user : ', envConfig.database.user);
       callback();
     });
   });
@@ -45,11 +58,11 @@ function createUser(callback) {
 function createDatabase(callback) {
   'use strict';
 
-  console.log('Creating database...');
+  console.log('Creating database : ', envConfig.database.database);
 
   connect(function(client, done) {
     var query = {
-      text: 'create database ' + dbConfig.database + ' with owner = ' + dbConfig.user,
+      text: 'CREATE DATABASE ' + envConfig.database.database + ' WITH OWNER = ' + envConfig.database.user
     };
 
     client.query(query, function(err, result) {
@@ -60,7 +73,8 @@ function createDatabase(callback) {
         return callback();
       }
 
-      console.log('Database created');
+      console.log('Created database : ', envConfig.database.database);
+
       callback();
     });
   });
@@ -94,13 +108,13 @@ function createSchema(callback) {
 module.exports = function() {
   'use strict';
 
-  oldConfig = JSON.parse(JSON.stringify(config));
-  config.database.user = 'postgres';
-  config.database.database = 'postgres';
-
+  // use the postgres master uname/pass so we can create the user and DB
+  currentConfig = JSON.parse(JSON.stringify(postgresUserConfig));
   createUser(function() {
     createDatabase(function() {
-      config = oldConfig;
+
+      // use the normal env specific config user/pass to do everything else.
+      currentConfig = JSON.parse(JSON.stringify(envConfig));
       createSchema(function() {
         console.log('Done');
         process.exit();

--- a/server/config/config.development.json
+++ b/server/config/config.development.json
@@ -1,6 +1,8 @@
 {
   "database": {
     "type": "postgres",
+    "postgres_user": "postgres",
+    "postgres_user_password": "",
     "user": "crypton_development_user",
     "password": "crypton_development_user_password",
     "database": "crypton_development",

--- a/server/config/config.test.json
+++ b/server/config/config.test.json
@@ -1,6 +1,8 @@
 {
   "database": {
     "type": "postgres",
+    "postgres_user": "postgres",
+    "postgres_user_password": "",
     "user": "crypton_test_user",
     "password": "crypton_test_user_password",
     "database": "crypton_test",

